src/game_events/handlers.cpp

```diff
@@ -241,10 +241,29 @@ struct filter_attack : public event_filter {
 	{
 		const unit_map& units = resources::gameboard->units();
 		const auto& loc = first_ ? event_info.loc1 : event_info.loc2;
-		auto unit = units.find(loc);
-		if(unit != units.end() && loc.matches_unit(unit)) {
-			const config& attack = event_info.data.child(first_ ? "first" : "second");
-			return swf_.empty() || matches_special_filter(attack, swf_);
+		const auto& loc_d = first_ ? event_info.loc2 : event_info.loc1;
+		auto unit_a = units.find(loc);
+		auto unit_d = units.find(loc_d);
+		if(unit_a != units.end() && loc.matches_unit(unit_a)) {
+			const unit& u = *unit_a;
+			const config& temp_weapon = event_info.data.child(first_ ? "first" : "second");
+			auto weapon = std::make_shared<const attack_type>(temp_weapon);
+			const_attack_ptr attack = weapon->shared_from_this();
+			if(unit_d != units.end() && loc_d.matches_unit(unit_d)) {
+				const unit& opp = *unit_d;
+				const config& temp_other_weapon = event_info.data.child(!first_ ? "first" : "second");
+				auto opp_weapon = std::make_shared<const attack_type>(temp_other_weapon);
+				const_attack_ptr second_attack = opp_weapon->shared_from_this();
+				auto ctx = attack->specials_context(u.shared_from_this(), opp.shared_from_this(), loc, loc_d, first_, second_attack);
+				std::optional<decltype(ctx)> opp_ctx;
+				if(second_attack) {
```

Comment:
IIUC, second_attack is never false here, so 3 lines can be reduced to just
`auto opp_ctx = opp_weapon->specials_context...`.

Reply:
and if second_unit is Dark Adept and attackis melee, second_attack will be nulll, isn't it?

Reply:
Ah, I added the edited paragraph just at the same time as you replied. It's not `null`: https://stackoverflow.com/questions/61095800/is-it-possible-that-make-shared-has-no-any-exception-but-returns-a-nullptr

Reply:
No need for std::optional.

Reply:
i prefer what it remain foruse emplace( instead of auto opp_ctx = opp_weapon->specials_context....

Reply:
Did some testing, and it seems the `opp_ctx` isn't needed at all. Same results with and without the extra bit of test that I've added in a separate comment.

* The test passes when the `auto ctx =` line is there, even without `opp_ctx`
* It fails when `ctx` is removed, even if the `opp_ctx` line is still there.

I tested with these two ways of declaring `opp_ctx` with `decltype(ctx)`:
```
std::optional<attack_type::specials_context_t> opp_ctx;
// followed by the unchanged opp_ctx.emplace(...) line
```

```
auto opp_ctx = second_attack->specials_context(opp.shared_from_this(), u.shared_from_this(), loc_d, loc, !first_, attack);
```

Reply:
There is one case where opp_ctx is still necessary, and that is when the activity of the special weapon being tested depends on the presence and activity of another special weapon held by the opponent. By testing the absence of opp_ctx in this configuration, nothing happens, the declaration of opp_ctx is changed, okay, I detect no problem, but this variable remains necessary and should not be removed.

Reply:
declaration changed, but opp_ctx remain.

Reply:
I'm wondering if that's handled by the `second_attack` argument in
```
auto ctx = attack->specials_context(u.shared_from_this(), opp.shared_from_this(), loc, loc_d, first_, second_attack);
```
instead of relying on `opp_ctx`.

Reply:
if opp_ctx unused, then second_attack has const config data only, for what fix will be perfect, both weapon must use ctx and opp_ctx.